File: Perl/day20.pl
#!/usr/bin/env perl
BEGIN {
    use Cwd;
    our $directory = cwd;
    our $local_lib = $ENV{"HOME"} . '/perl5/lib/perl5';
}

use lib $directory;
use lib $local_lib;

use Modern::Perl 2022;
use autodie;
use Data::Dumper;
#use Storable 'dclone';

use AOC::Util;
#use AOC::SpatialUtil;

my $INPUT_PATH = '../input';
#my $INPUT_FILE = 'day20_test.txt';
my $INPUT_FILE = 'day20_challenge.txt';
my @input = read_input("$INPUT_PATH/$INPUT_FILE");

say "Advent of Code 2022, Day 20: Grove Positioning System";

my @linked_list = build_dll(@input);

my $p1 = solve(\@linked_list, 1);
say "Part One: the sum of the coord values is $p1.";

@linked_list = build_dll(@input);
for my $i (0..$#linked_list) {
	$linked_list[$i]{'value'} *= 811589153;
}
my $p2 = solve(\@linked_list, 10);
say "Part Two: the sum of the coord values is $p2.";

exit( 0 );

sub solve {
	my ($dll, $mix_count) = @_;
	
	my $zero;
	for my $rep (1..$mix_count) {
		$zero = mix_dll($dll);
	}
	
	my @coord = ();
	my $ptr = $zero;
	
	for my $i (1..3000) {
		$ptr = $ptr->{'next'};
		if ($i % 1000 == 0) { push(@coord, $ptr->{'value'}); }
	}
	
	say join(', ', @coord);
	my $sum = 0;
	for my $c (@coord) {
		$sum += $c;
	}
	return $sum;
}

sub build_dll {
	my @input = @_;
	my @dll = ();
	for my $line (@input) {
		push(@dll, {'value' => int($line)});
	}
	for my $i (0..$#dll) {
		my $prev = $i-1;
		my $next = $i+1;
		$prev = $#dll if ($prev < 0);
		$next = 0 if ($next >$#dll);
		$dll[$i]{'prev'} = $dll[$prev];
		$dll[$i]{'next'} = $dll[$next];
	}
	return @dll;
}

sub mix_dll {
	my $dll = shift;
	my $zero;
	my $list_length = scalar( @{$dll} ) -1;
	
	for my $i (0..$#{$dll}) {
		my $mover = $dll->[$i];
		if ($mover->{'value'} == 0) { $zero = $mover };
		
		my $diff = abs($mover->{'value'})  % $list_length;
		if ($diff > 0) {
			# Remove links to $mover
			$mover->{'prev'}{'next'} = $mover->{'next'};
			$mover->{'next'}{'prev'} = $mover->{'prev'};
			
			my $dest_prev = $mover;
			if ($mover->{'value'} > 0) {
				for my $j (1..$diff) { $dest_prev = $dest_prev->{'next'}; }
			}
			elsif ($mover->{'value'} < 0) {
				for my $j (1..$diff+1) { $dest_prev = $dest_prev->{'prev'}; }
			}
			my $dest_next = $dest_prev->{'next'};
		
			# Insert $mover
			$dest_prev->{'next'} = $mover;
			$dest_next->{'prev'} = $mover;
			$mover->{'prev'} = $dest_prev;
			$mover->{'next'} = $dest_next;
		}
	}	
	#print_dll($dll, $zero);
	
	return $zero;
}

sub print_dll {
	my ($dll, $first) = @_;
	my @values = ();
	my $ptr = $first;
	for my $i (0..$#{$dll}) {
		push(@values, $ptr->{'value'});
		$ptr = $ptr->{'next'};
	}
	say join(', ', @values);
}

File: Swift/AoC 2022/AoC 2022/Solutions/Day20.swift
//
//  Day20.swift
//  AoC 2022
//
//  Created by Simon Biickert on 2022-12-20.
//

import Foundation

class Day20: AoCSolution {
	override init() {
		super.init()
		day = 20
		name = "Grove Positioning System"
	}
	
	override func solve(filename: String, index: Int) -> AoCResult {
		super.solve(filename: filename, index: index)
		
		let input = AoCUtil.readInputFile(named: filename, removingEmptyLines: true)
		
		var dll = DoubleLinkedList(input)
		let p1 = solvePart(linkedList: dll, repeat: 1)
		print("Part One: the sum of the coord values is \(p1).")
		
		dll = DoubleLinkedList(input)
		for node in dll.nodes {
			node.value *= 811589153;
		}
		let p2 = solvePart(linkedList: dll, repeat: 10)
		print("Part Two: the sum of the coord values is \(p2).")
		
		return AoCResult(part1: String(p1), part2: String(p2))
	}
	
	private func solvePart(linkedList dll: DoubleLinkedList, repeat count: Int) -> Int {
		for _ in (1...count) {
			mix(dll)
		}
		
		var coord = [Int]()
		var ptr = dll.zero
		
		for i in 1...3000 {
			ptr = ptr!.next
			if (i % 1000 == 0) { coord.append(ptr!.value)}
		}
		print(coord.map({ String($0) }).joined(separator: ", "))
		let sum = coord.reduce(0, +)
		return sum
	}
	
	private func mix(_ dll: DoubleLinkedList) {
		for i in 0..<dll.nodes.count {
			let mover = dll.nodes[i]
			let diff = abs(mover.value) % (dll.nodes.count - 1)
			dll.move(node: mover, by: diff)
		}
	}
}

class DoubleLinkedList {
	var nodes = [DLLNode]()
	var zero: DLLNode?
	
	init(_ input: [String]) {
		for line in input {
			nodes.append(DLLNode(value: Int(line)!))
		}
		for i in (0..<nodes.count) {
			let prev = i > 0 ? i - 1 : nodes.count - 1
			let next = i < nodes.count-1 ? i + 1 : 0
			nodes[i].prev = nodes[prev]
			nodes[i].next = nodes[next]
			if nodes[i].value == 0 {
				zero = nodes[i]
			}
		}
	}
	
	func move(node mover: DLLNode, by diff: Int) {
		if (diff > 0) {
			// Remove links to mover
			mover.prev!.next = mover.next
			mover.next!.prev = mover.prev

			var dest_prev = mover
			if (mover.value > 0) {
				for _ in 1...diff {
					dest_prev = dest_prev.next!
				}
			}
			else if (mover.value < 0) {
				for _ in 1...diff+1 {
					dest_prev = dest_prev.prev!
				}
			}
			let dest_next = dest_prev.next!

			// Insert mover
			dest_prev.next = mover
			dest_next.prev = mover
			mover.prev = dest_prev
			mover.next = dest_next
		}
	}
	
	func toString() -> String {
		var ptr = zero
		var values = [String]()
		for _ in 0..<nodes.count {
			values.append(String(ptr!.value))
			ptr = ptr!.next
		}
		return values.joined(separator: ", ")
	}
}

class DLLNode {
	var value: Int
	var prev: DLLNode?
	var next: DLLNode?
	
	init(value: Int) {
		self.value = value
	}
}
